run/src/types.ts

```diff
@@ -18,36 +18,42 @@ export interface DevToolEnabledSource {
   _isCycleSource: string;
 }
 
-export type Sources = {
-  [name: string]: any;
-};
-
-export type Sinks = {
-  [name: string]: any;
-};
-
-export type SinkProxies<Si extends Sinks> = {[P in keyof Si]: Stream<any>};
+export type SinkProxies<Si> = {[P in keyof Si]: Stream<any>};
 
-export type FantasySinks<Si> = {[S in keyof Si]: FantasyObservable};
-
-export interface Driver<Sink, Source> {
-  (stream: Sink, driverName?: string): Source;
+export interface Driver<Si, So> {
+  (stream: Si, driverName?: string): So;
 }
 
-export type Drivers<So extends Sources, Si extends Sinks> = {
-  [P in keyof (So & Si)]: Driver<Si[P], So[P]>
-};
-
 export type DisposeFunction = () => void;
 
-export interface CycleProgram<So extends Sources, Si extends Sinks> {
-  sources: So;
-  sinks: Si;
+export type Drivers = {
+  [name: string]: (...args: Array<any>) => any;
+};
+export type Main = (...args: Array<any>) => any;
+export type Sources<D extends Drivers> = {[k in keyof D]: ReturnType<D[k]>};
+export type Sinks<M extends Main> = {
+  [k in (string & keyof ReturnType<M>)]: ReturnType<M>[k] & FantasyObservable
+};
+
+export interface CycleProgram<D extends Drivers, M extends Main> {
+  sources: Sources<D>;
+  sinks: Sinks<M>;
   run(): DisposeFunction;
 }
 
-export interface Engine<So extends Sources, Si extends Sinks> {
-  sources: So;
-  run(sinks: Si): DisposeFunction;
+export interface Engine<D extends Drivers> {
+  sources: Sources<D>;
+  run<M extends Main>(sinks: Sinks<M>): DisposeFunction;
   dispose(): void;
 }
+
+export type MatchingMain<D extends Drivers, M extends Main> = Main & {
+  (so?: Partial<Sources<D>>): Sinks<M>;
```

Comment:
What exactly does this do?

Reply:
MatchingMain is basically the "set of main functions that can possibly match the given drivers object" and MatchingDrivers is the "set of driver objects that can possibly match the given main function".